c_glib/plasma-glib/client.cpp

```diff
@@ -21,47 +21,192 @@
+{
+  auto priv = GPLASMA_CLIENT_CREATE_OPTIONS_GET_PRIVATE(object);
+
+  switch (prop_id) {
+  case PROP_GPU_DEVICE:
+    g_value_set_int(value, priv->gpu_device);
+    break;
+  default:
+    G_OBJECT_WARN_INVALID_PROPERTY_ID(object, prop_id, pspec);
+    break;
+  }
+}
+
+static void
+gplasma_client_create_options_init(GPlasmaClientCreateOptions *object)
+{
+}
+
+static void
+gplasma_client_create_options_class_init(GPlasmaClientCreateOptionsClass *klass)
+{
+  auto gobject_class = G_OBJECT_CLASS(klass);
+
+  gobject_class->set_property = gplasma_client_create_options_set_property;
+  gobject_class->get_property = gplasma_client_create_options_get_property;
+
+  GParamSpec *spec;
+  spec = g_param_spec_int("gpu-device",
+                          "GPU device",
+                          "The GPU device number. -1 means GPU isn't used.",
+                          -1,
+                          G_MAXINT,
+                          -1,
+                          static_cast<GParamFlags>(G_PARAM_READWRITE |
+                                                   G_PARAM_CONSTRUCT));
+  g_object_class_install_property(gobject_class, PROP_GPU_DEVICE, spec);
+}
+
+/**
+ * gplasma_client_create_options_new:
+ *
+ * Returns: A newly created #GPlasmaClientCreateOptions.
+ *
+ * Since: 0.12.0
+ */
+GPlasmaClientCreateOptions *
+gplasma_client_create_options_new(void)
+{
+  auto options = g_object_new(GPLASMA_TYPE_CLIENT_CREATE_OPTIONS,
+                              NULL);
+  return GPLASMA_CLIENT_CREATE_OPTIONS(options);
+}
+
+/**
+ * gplasma_client_create_options_set_metadata:
+ * @options: A #GPlasmaClientCreateOptions.
+ * @metadata: (nullable) (array length=size): The metadata of a created object.
+ * @size: The number of bytes of the metadata.
+ *
+ * Since: 0.12.0
+ */
+void
+gplasma_client_create_options_set_metadata(GPlasmaClientCreateOptions *options,
+                                           const guint8 *metadata,
+                                           gsize size)
+{
+  auto priv = GPLASMA_CLIENT_CREATE_OPTIONS_GET_PRIVATE(options);
+  if (priv->metadata) {
+    g_free(priv->metadata);
+  }
+  priv->metadata = static_cast<guint8 *>(g_memdup(metadata, size));
+  priv->metadata_size = size;
+}
+
+/**
+ * gplasma_client_create_options_get_metadata:
+ * @options: A #GPlasmaClientCreateOptions.
+ * @size: (nullable) (out): The number of bytes of the metadata.
+ *
+ * Returns: (nullable) (array length=size): The metadata of a created object.
+ *
+ * Since: 0.12.0
+ */
+const guint8 *
+gplasma_client_create_options_get_metadata(GPlasmaClientCreateOptions *options,
+                                           gsize *size)
+{
+  auto priv = GPLASMA_CLIENT_CREATE_OPTIONS_GET_PRIVATE(options);
+  if (size) {
+    *size = priv->metadata_size;
+  }
+  return priv->metadata;
+}
+
 typedef struct GPlasmaClientPrivate_ {
-  std::shared_ptr<plasma::PlasmaClient> client;
+  plasma::PlasmaClient *client;
```

Comment:
Is that because this change prevent a memory leak?

Reply:
No.
`shared_ptr` is just not needed here. We don't need to share a `plasma::PlasmaClient` created by `plasma_client_new()`.

Reply:
I see.